### DISCUSSION TITLE
How to perform a search with `useInfiniteQuery`

### DISCUSSION THREAD

**Discussion by jorisre** [blitz-js/blitz#2128] [Q&A] [answered]
👋🏻 

I'd like to perform a search with `useInfiniteQuery` hook.
My home page has 2 behaviors:
- home: it's just a list of last items
- search input: when the user fill the search input, I want to perform a search and update/replace the items list

![Screen Shot 2021-03-19 at 16 37 14](https://user-images.githubusercontent.com/7545547/111807422-3561a200-88d3-11eb-8162-337ef2ee6e76.png)

### Example:

```tsx
function App() {
 const [search, setSearch] = useState('');
 
  const where = search ? { where: { name: { contains: search } } } : undefined; // Doesn't work
 
  const [
    groupedProjects,
    { isFetchingMore, fetchMore, canFetchMore, refetch, query },
  ] = useInfiniteQuery(getProjects, (page = { take: ITEMS_PER_PAGE, skip: 0, where }) => page, {
    getFetchMore: (lastGroup) => lastGroup.nextPage,
  })

  return (
    <div>
        <input type="search" value={search} onChange={() => {
           setSearch(e.currentTarget.value)
        }} />
        {groupedProjects.map((group, i) => (
            <Fragment key={i}>
                {group.projects.map((project) => (
                    <div key={project.id}>{project.name}</div>
                ))}
            </Fragment>
         )}
   </div>
  )
}
```

How can I pass a new key to `useInfiniteQuery` to make a new request ?  it's the default behavior with react-query when a key change but blitz do it for us

Thanks for your help 🙏🏻

**Comment by flybayer** [comment]
That's a really good question! Not sure if it's currently possible 🤔 

I'll ask some others, but could you go ahead and open this as a feature/change issue?

**Reply by jorisre** [reply]
Thanks @flybayer , is there a way to setup manually the react-query key ?

**Reply by flybayer** [reply]
Not currently! But we could also add that

**Comment by flybayer** [comment, marked as answer]
Update! This should work right now. Change your second `useInifiniteQuery` to put `where` in the return value
```diff
-(page = { take: ITEMS_PER_PAGE, skip: 0, where }) => page
+(page = { take: ITEMS_PER_PAGE, skip: 0 }) => ({...page, where})
```

**Reply by jorisre** [reply]
It works on first render but it doesn't work when I change dynamically the `where` condition. It doesn't trigger a new request when I change it dynamically.
_(when the search input's value changes in my case)_

To my knowledge, the only way to achieve that is to change the react-query key

**Reply by flybayer** [reply]
We use the result of that function (`({...page, where})`) to compute the query key, so theoretically it should work.

Can you create a [codesandbox](https://sandbox.blitzjs.com) or github repo with a minimal reproduction of this? And then open a bug report for this so it doesn't get lost.

**Reply by jorisre** [reply]
Yes I'll, thanks for the help

**Reply by jorisre** [reply]
Issue update:

It just doesn't work when I use `initialData` of `useInfiniteQuery`. Everything works well when I don't use `initialData` with or without `suspense`

Here is the minimal reproduction repo https://github.com/jorisre/blitzjs-search-with-react-query


1. Clone the repo, install deps
2. Run seeds: `blitz db seed`
3. Start the project `yarn dev`
4. Create an account
5. Go to http://localhost:3000/projects and try to search a person by name _(nothing happen)_. Remove one char from the search input and the request is triggered

ℹ️ `suspense` is disabled and `useInfiniteQuery` have `initialData` from `getServerSideProps`. If `initialData` commented, the search works well

![blitz-ss](https://user-images.githubusercontent.com/7545547/112736952-faa4dd00-8f56-11eb-8606-4a62210e94f7.gif)

**Reply by flybayer** [reply]
@jorisre ok, can you please open an issue with those details so we can track it properly?
